icechunk/src/storage/s3.rs

```diff
@@ -126,17 +134,25 @@ pub async fn mk_client(config: Option<&S3Config>) -> Client {
 }
 
 impl S3Storage {
-    pub async fn new_s3_store(
-        bucket_name: impl Into<String>,
-        prefix: impl Into<String>,
-        config: Option<&S3Config>,
-    ) -> Result<S3Storage, StorageError> {
-        let client = Arc::new(mk_client(config).await);
-        Ok(S3Storage { client, prefix: prefix.into(), bucket: bucket_name.into() })
+    pub async fn new_s3_store(config: &S3Config) -> Result<S3Storage, StorageError> {
+        let client =
+            RwLock::new(Some(Arc::new(mk_client(config.options.as_ref()).await)));
+        Ok(S3Storage { client, config: config.clone() })
+    }
+
+    async fn get_client(&self) -> Arc<Client> {
+        let guard = self.client.read().await;
+        if let Some(client) = guard.as_ref() {
+            Arc::clone(client)
+        } else {
+            let client = Arc::new(mk_client(self.config.options.as_ref()).await);
+            *self.client.write().await = Some(Arc::clone(&client));
```

Comment:
why does this work? How can you lock for write if you are still holding to a read lock?

Reply:
I have no idea, i will make it temporary though just in case